tests/deploy_test.go

```diff
@@ -5,121 +5,99 @@ import (
 	"fmt"
 	"net/http"
 	"sort"
-	"strings"
 	"testing"
 
 	sdk "github.com/openfaas/faas-cli/proxy"
 )
 
 var emptyQueryString = ""
 
-func Test_Deploy_Stronghash(t *testing.T) {
-	envVars := map[string]string{}
-	functionRequest := &sdk.DeployFunctionSpec{
-		Image:        "functions/alpine:latest",
-		FunctionName: "stronghash",
-		Network:      "func_functions",
-		FProcess:     "sha512sum",
-		EnvVars:      envVars,
-	}
-
-	deployStatus := deploy(t, functionRequest)
-	if deployStatus != http.StatusOK && deployStatus != http.StatusAccepted {
-		t.Fatalf("got %d, wanted %d or %d", deployStatus, http.StatusOK, http.StatusAccepted)
-	}
-
-	list(t, http.StatusOK)
+type FunctionMetaSchema struct {
+	Image        string
+	FunctionName string
+	Network      string
+	FProcess     string
+	EnvVars      map[string]string
+	Annotations  map[string]string
+	Labels       map[string]string
+	Namespace    string
 }
 
-func Test_Deploy_PassingCustomEnvVars_AndQueryString(t *testing.T) {
-	envVars := map[string]string{}
-	envVars["custom_env"] = "custom_env_value"
-
-	functionRequest := &sdk.DeployFunctionSpec{
-		Image:        "functions/alpine:latest",
-		FunctionName: "env-test",
-		Network:      "func_functions",
-		FProcess:     "env",
-		EnvVars:      envVars,
-	}
+func Test_Deploy_MetaData(t *testing.T) {
+	cases := []FunctionMetaSchema{
```

Comment:
this might be slightly better if we did this

```go
cases := []struct{
   name string
   function FunctionMetaSchema
}{
   {
      name: "deploy without any extra metadata",
      function: {
			Image:        "functions/alpine:latest",
			FunctionName: "stronghash",
			Network:      "func_functions",
			FProcess:     "sha512sum",
			EnvVars:      map[string]string{},
			Annotations:  map[string]string{},
			Labels:       map[string]string{},
			Namespace:    config.DefaultNamespace,
		},
   },
   {
       name: "deploy with labels",
   },
   {
       name: "deploy with annotations",
   },
   // and so on
}
```

Obviously line 71 would then look like 

```go
c.name = fmt.Sprintf("%s to %s", c.name, config.Namespaces[0])
c.function.Namespace = config.Namespaces[0]
```

Then inside the loop over cases we can wrap it in sub-tests

```go
for _, c := range cases {
    t.Run(c.name, func(t *testing.T) { 
        // you existing test logic here 
    })
}
```


The main benefit to this tweak is that we can add context to each test case which makes it clear why it is there and if it fails, what the exact feature that is failing in that provider, for example, "function labels are broken" etc.  It will also help distinguish if it is broken for all namespaces or just the additional namespaces.

Reply:
Thanks for review. This really improves the code.